apiserver/facades/agent/uniter/networkinfo.go

```diff
@@ -422,20 +411,142 @@ func (n *NetworkInfo) NetworksForRelation(
 
 // machineNetworkInfos returns network info for the unit's machine based on
 // devices with addresses in the input spaces.
-// TODO (manadart 2019-10-10): `GetNetworkInfoForSpaces` is only used here and
-// could be relocated from the state package, reducing cross-cutting concerns
-// there.
-func (n *NetworkInfo) machineNetworkInfos(spaces ...string) (map[string]state.MachineNetworkInfoResult, error) {
+func (n *NetworkInfo) machineNetworkInfos(spaceIDs ...string) (map[string]machineNetworkInfoResult, error) {
 	machineID, err := n.unit.AssignedMachineId()
 	if err != nil {
-		return nil, err
+		return nil, errors.Trace(err)
 	}
 	machine, err := n.st.Machine(machineID)
 	if err != nil {
-		return nil, err
+		return nil, errors.Trace(err)
+	}
+
+	spaceSet := set.NewStrings(spaceIDs...)
+
+	results := make(map[string]machineNetworkInfoResult)
+
+	var privateIPAddress string
+
+	if spaceSet.Contains(corenetwork.AlphaSpaceId) {
+		var err error
+		privateMachineAddress, err := n.pollForAddress(machine.PrivateAddress)
+		if err != nil {
+			results[corenetwork.AlphaSpaceId] = machineNetworkInfoResult{Error: errors.Annotatef(
+				err, "getting machine %q preferred private address", machine.MachineTag())}
+
+			// Remove this ID to prevent further processing.
+			spaceSet.Remove(corenetwork.AlphaSpaceId)
+		} else {
+			privateIPAddress = privateMachineAddress.Value
+		}
 	}
 
-	return machine.GetNetworkInfoForSpaces(set.NewStrings(spaces...)), nil
+	// Link-layer devices are set in a single transaction for all devices
+	// observed on the machine, so the first result will include them all.
+	var addresses []*state.Address
+	retryArg := n.retryFactory()
+	retryArg.Func = func() error {
+		var err error
+		addresses, err = machine.AllAddresses()
+		return err
+	}
+	retryArg.IsFatalError = func(err error) bool {
+		return err != nil
+	}
+	if err := retry.Call(retryArg); err != nil {
+		result := machineNetworkInfoResult{Error: errors.Annotate(err, "getting devices addresses")}
+		for _, id := range spaceSet.Values() {
+			if _, ok := results[id]; !ok {
+				results[id] = result
+			}
+		}
+		return results, nil
+	}
+
+	logger.Debugf("Looking for address from %v in spaces %v", addresses, spaceIDs)
+
+	var privateLinkLayerAddress *state.Address
+	for _, addr := range addresses {
+		subnet, err := addr.Subnet()
+		switch {
```

Comment:
How come did you opt for a switch instead of an if block + continue for error handling?

Reply:
As mentioned, I don't specifically recall why it was like that prior - it was moved here from `state` with the lightest possible touch.